File: backtracking尾递归/17. Letter Combinations of a Phone Number/Solution.py
import itertools

class Solution:
    def __init__(self):
        self.dict = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
    def letterCombinations(self, digits):
        if len(digits) < 1:
            return []
        # 例如23， 想要的数据结构， [ [a,b,c], [d,e,f]]
        cases = [[i for i in self.dict[digit]] for digit in digits]
        # 返回 ['ab','cd', ... ]
        return [''.join(i) for i in itertools.product(*cases)]
    
    # 使用队列来解决问题
    # 空间上，使用一个队列，占据大概3的n次方，时间上则是3n
    def letterCombinationsUsingQueue(self, digits):
        if len(digits) < 1:
            return []
        # initialize
        queue = [i for i in self.dict[digits[0]]]
        for i, v in enumerate(digits[1:]):
            # when the head element's length does not meet, we need to extend it by all the possible alphabets, 
            # then push the result list back to queue in the end
            while len(queue[0]) < i + 2:
                material = queue.pop(0)
                new_add = [material + m for m in self.dict[v]]
                queue.extend(new_add)
        return queue
        

if __name__ == "__main__":
   s = Solution()
#    result = s.letterCombinations("245")
   result = s.letterCombinationsUsingQueue("245")
   print(result, len(result))

File: backtracking尾递归/17. Letter Combinations of a Phone Number/Solution.java
import java.util.ArrayList;


// TODO need to include both backtracking(dfs) and iterative(bfs) solution
class Solution{
    public static void main(String[] args) {
        String digits = "23";
        List<String> ret = letterCombinations_d(digits);
        System.out.println(ret);
    }

    private static final String[] KEYS = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
    public List<String> letterCombinations_d(String digits) {
        List<String> ret = new ArrayList<>();
        if (digits == null || digits.length() < 1){
            return ret;
        }
        combination("", digits, 0, ret);
        return ret;
    }

    // 空间上3(0)+3(1)+...+3(n-1)? 时间复杂度为3的n次方？
    private void combination(String buffer, String digits, int offset, List<String> ret){
        if(offset >= digits.length()){
            ret.add(buffer);
            return;
        }
        String letters = KEYS[(digits.charAt(offset) - '0')];
        for ( int i = 0 ; i < letters.length(); i++){
            combination(buffer + letters.charAt(i), digits, offset + 1, ret);
        }

    }
}